### PR TITLE
RabbitMQ 3.7.9+ list compatibility and provider cleanup

### LINKED ISSUE #740: rabbitmqctl 'broken' in 3.7.9
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 4.10.5
- Ruby: 2.1.9p490
- Distribution: CentOS 7
- Module version: v8.4.0

## How to reproduce (e.g Puppet code you use)

```
class { 'rabbitmq':
  delete_guest_user => true,
}
```

## What are you seeing

Error: Failed to apply catalog: Cannot parse invalid user line: user    tags

## What behaviour did you expect instead

No error

## Output log

Debug: Executing: '/usr/sbin/rabbitmqctl -q list_users'

This is generating since 3.7.9:

```
user    tags
guest   [administrator]
```

The first line should be removed...

## Any additional information you'd like to impart

See: https://github.com/rabbitmq/rabbitmq-cli/issues/273

Seems a '--no-table-headers'  needs to be added...

Too bad this parameter is not supported in 3.7.8-1 and (likely) lower...

### PR DESCRIPTION
**Pull request by JayH5**
#### Pull Request (PR) description

This is an alternative attempt at fixing `rabbitmqctl list_{x}` commands with RabbitMQ 3.7.9+. #751 was also an attempt. This PR takes a different approach by adding a class method in the "parent" provider class for all the providers. This method provides a consistent way to run `rabbitmqctl list_` operations in all provider classes.

This PR also cleans up the process of finding RabbitMQ CLI binaries: all providers now share the same command methods. It also adds caching to the RabbitMQ version value.

#### This Pull Request (PR) fixes the following issues

Fixes #740

### CODE REVIEW

File: lib/puppet/provider/rabbitmq_cli.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      if Puppet::Util::Package.versioncmp(rabbitmq_version, '3.7.9') >= 0
+        ['-q', '--no-table-headers']
+      else
+        ['-q']
+      end
+    rabbitmqctl("list_#{resource}", *list_opts, *opts)
+  end
+
+  # Retry the given code block 'count' retries or until the
+  # command suceeeds. Use 'step' delay between retries.
```
Comment (alexjfisher) [inline lib/puppet/provider/rabbitmq_cli.rb]: Spelling of ‘succeeds’

**Comment by wyardley** [comment]
Thanks for the contribution @JayH5 
I know a ton of people will appreciate this
I squash-merged, since the commits weren't signed anyway; hope that's Ok.

**Comment by JayH5** [comment]
Thanks!
> I squash-merged, since the commits weren't signed anyway; hope that's Ok.

Yeah that's fine. Would've done that myself but just pushed this PR as I was ending my work day.

**Comment by wyardley** [comment]
Great! We are trying to cut a release of this soon.

**Pull request merged**